File: Python/Combination Sum II.py
class Solution(object):
    """
    :type candidates: List[int]
    :type target: int
    :rtype: List[List[int]]
    """
    def combinationSum2(self, cans, t):
        if cans == None or len(cans) == 0 or t <= 0:
            return []
            
        res = []
        cans.sort()
        self.find(cans, t, 0, [], res)
        return res
        
    def find(self, cans, t, index, path, res):
        if t == 0:
            res.append(list(path))
            return
        
        pre = -1
        for i in range(index, len(cans), 1):
            if cans[i] > t:
                break
            if pre != -1 and pre == cans[i]:
                continue
            
            path.append(cans[i])
            self.find(cans, t-cans[i], i+1, path, res)
            del path[len(path)-1]
            
            pre = cans[i]
            

File: Java/Combination Sum II.java
public class Solution {
    public List<List<Integer>> combinationSum2(int[] cans, int t) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        if (cans == null || cans == null || t <= 0)
            return res;
            
        Arrays.sort(cans);
        find(cans, t, 0, new ArrayList<Integer>(), res);
        return res;
    }
    
    private void find(int[] cans, int t, int index, ArrayList<Integer> path, List<List<Integer>> res) {
        if (t == 0) {
            res.add(new ArrayList<Integer>(path));
            return;
        }
        
        int pre = -1;
        for (int i=index; i<cans.length; i++) {
            if (cans[i] > t) break;
            if (pre != -1 && pre == cans[i]) continue;
            
            path.add(cans[i]);
            find(cans, t-cans[i], i+1, path, res);
            path.remove(path.size()-1);
            
            pre = cans[i];
        }
    }
}